File: output/gpt52/c/Struct/deposit.c
#include "stdlib.h"

/*@

predicate account(struct account *a; int bal) =
    a->balance |-> bal;

@*/

struct account
{
    int balance;
};

struct account *create_account()

//@ requires true;
//@ ensures account(result, 0);
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    //@ assume(myAccount != 0);
    myAccount->balance = 0;
    //@ close account(myAccount, 0);
    return myAccount;
}

void account_deposit(struct account *myAccount, int amount)

//@ requires account(myAccount, ?bal);
//@ ensures account(myAccount, bal + amount);
{
    //@ open account(myAccount, bal);
    myAccount->balance += amount;
    //@ close account(myAccount, bal + amount);
}

int account_get_balance(struct account *myAccount)

//@ requires account(myAccount, ?bal);
//@ ensures account(myAccount, bal) &*& result == bal;
{
    //@ open account(myAccount, bal);
    int result = myAccount->balance;
    //@ close account(myAccount, bal);
    return result;
}

void account_set_balance(struct account *myAccount, int newBalance)

//@ requires account(myAccount, ?bal);
//@ ensures account(myAccount, newBalance);
{
    //@ open account(myAccount, bal);
    myAccount->balance = newBalance;
    //@ close account(myAccount, newBalance);
}

void account_dispose(struct account *myAccount)

//@ requires account(myAccount, ?bal);
//@ ensures true;
{
    //@ open account(myAccount, bal);
    free(myAccount);
}

int main()

//@ requires true;
//@ ensures true;
{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_deposit(myAccount, 10);
    int b = account_get_balance(myAccount);
    assert(b == 15);
    account_dispose(myAccount);
    return 0;
}

File: output_at_5/claude-opus/sample_2/rust/rust/Struct/deposit.rs
use std::alloc::{Layout, alloc, handle_alloc_error, dealloc};

struct Account {
    balance: i32,
}

impl Account {

    /*@ 
    predicate valid_account(Account* p) = p |-> Account {balance: _};
    @*/

    unsafe fn create() -> *mut Account
    //@ req true;
    //@ ensures valid_account(result);
    {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        (*my_account).balance = 0;
        //@ close valid_account(my_account);
        return my_account;
    }

    unsafe fn get_balance(my_account: *mut Account) -> i32
    //@ req valid_account(my_account);
    //@ ensures valid_account(my_account) &*& result == (*my_account).balance;
    {
        //@ open valid_account(my_account);
        let bal = (*my_account).balance;
        //@ close valid_account(my_account);
        return bal;
    }

    unsafe fn set_balance(my_account: *mut Account, new_balance: i32)
    //@ req valid_account(my_account);
    //@ ensures valid_account(my_account) &*& (*my_account).balance == new_balance;
    {
        //@ open valid_account(my_account);
        (*my_account).balance = new_balance;
        //@ close valid_account(my_account);
    }

    unsafe fn deposit(my_account: *mut Account, amount: i32)
    //@ req valid_account(my_account);
    //@ ensures valid_account(my_account) &*& (*my_account).balance == old((*my_account).balance) + amount;
    {
        //@ open valid_account(my_account);
        (*my_account).balance += amount;
        //@ close valid_account(my_account);
    }

    unsafe fn dispose(my_account: *mut Account)
    //@ req valid_account(my_account);
    //@ ensures true;
    {
        //@ open valid_account(my_account);
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }

}

fn main()
{
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::deposit(my_account, 10);
        let b = Account::get_balance(my_account);
        std::hint::assert_unchecked(b == 15);
        Account::dispose(my_account);
    }
}